commands/proxy.ts

```diff
@@ -0,0 +1,137 @@
+import * as commandParams from "../command-params";
+import { isInteractive } from "../helpers";
+import { EOL } from "os";
+
+const proxyGetCommandName = "proxy|*get";
+const proxySetCommandName = "proxy|set";
+const proxyClearCommandName = "proxy|clear";
+
+export abstract class ProxyCommandBase implements ICommand {
+	public disableAnalytics = true;
+	public allowedParameters: ICommandParameter[] = [];
+
+	constructor(protected $analyticsService: IAnalyticsService,
+		protected $logger: ILogger,
+		protected $proxyService: IProxyService,
+		private commandName: string) {
+	}
+
+	public abstract execute(args: string[]): Promise<void>;
+
+	protected async tryTrackUsage() {
+		try {
+			await this.$analyticsService.trackFeature(this.commandName);
+		} catch (ex) {
+			this.$logger.trace("Error in trying to track proxy command usage:");
+			this.$logger.trace(ex);
+		}
+	}
+}
+
+export class ProxySetCommand extends ProxyCommandBase {
+	public allowedParameters = [new commandParams.StringCommandParameter(this.$injector), new commandParams.StringCommandParameter(this.$injector), new commandParams.StringCommandParameter(this.$injector), new commandParams.StringCommandParameter(this.$injector)];
+
+	constructor(private $errors: IErrors,
+		private $injector: IInjector,
+		private $prompter: IPrompter,
+		protected $analyticsService: IAnalyticsService,
+		protected $logger: ILogger,
+		protected $proxyService: IProxyService) {
+		super($analyticsService, $logger, $proxyService, proxySetCommandName);
+	}
+
+	public async execute(args: string[]): Promise<void> {
+		let hostname = args[0],
+			port = args[1],
+			username = args[2],
+			password = args[3];
+
+		const noHostName = !hostname;
+		const noPort = !port;
+		const credentialsRequired = username && !password;
+
+		if (!isInteractive() && (noHostName || noPort || credentialsRequired)) {
+			this.$errors.fail("Console is not interactive - you need to supply all command parameters.");
+		}
+
+		if (noHostName) {
+			hostname = await this.$prompter.getString("Hostname", { allowEmpty: false });
+		}
+
+		if (noPort) {
+			port = await this.$prompter.getString("Port", { allowEmpty: false });
+		}
+
+		if (!username) {
+			username = await this.$prompter.getString("Username (Leave empty to omit.)", { defaultAction: () => "" });
+		}
+
+		if (credentialsRequired) {
+			password = await this.$prompter.getPassword("Password");
+		}
+
+		if (username && password) {
+			await this.$proxyService.setCredentials({
+				password,
+				username
+			});
+		}
+
+		const proxyCache: IProxyCache = {
+			PROXY_HOSTNAME: hostname,
+			PROXY_PORT: +port,
+			USE_PROXY: true
+		};
+
+		this.$proxyService.setCache(proxyCache);
+		this.$logger.out("Sucessfully setup proxy.");
+		await this.tryTrackUsage();
+	}
+}
+
+$injector.registerCommand(proxySetCommandName, ProxySetCommand);
+
+export class ProxyGetCommand extends ProxyCommandBase {
```

Comment:
I guess this is the common practice, but when I navigate to classes if there are more than one in a file, it doesn't work correctly. What is more as a whole I've been used to the 1 file = 1 class rule.